src/dmd/astbase.d

```diff
@@ -3995,13 +4007,165 @@ struct ASTBase
             t.isproperty = isproperty;
             t.isref = isref;
             t.isreturn = isreturn;
-            t.isscope = isscope;
-            t.iswild = iswild;
+            t.isthisscope = isthisscope;
+            t.isreturninferred = isreturninferred;
+            t.isscopeinferred = isscopeinferred;
+            t.isInOutParam = isInOutParam;
+            t.isInOutQual = isInOutQual;
             t.trust = trust;
             t.fargs = fargs;
             return t;
         }
 
+        /// set or get if the function has the `nothrow` attribute
+        bool isnothrow() const pure
+        {
+            return (funcFlags & FunctionFlag.isnothrow) != 0;
+        }
+        /// ditto
+        void isnothrow(bool v)
+        {
+            if (v) funcFlags |= FunctionFlag.isnothrow;
+            else funcFlags &= ~FunctionFlag.isnothrow;
+        }
+
+        /// set or get if the function has the `@nogc` attribute
+        bool isnogc() const pure
+        {
+            return (funcFlags & FunctionFlag.isnogc) != 0;
+        }
+        /// ditto
+        void isnogc(bool v)
+        {
+            if (v) funcFlags |= FunctionFlag.isnogc;
+            else funcFlags &= ~FunctionFlag.isnogc;
+        }
+
+        /// set or get if the function has the `@property` attribute
+        bool isproperty() const pure
+        {
+            return (funcFlags & FunctionFlag.isproperty) != 0;
+        }
+        /// ditto
+        void isproperty(bool v)
+        {
+            if (v) funcFlags |= FunctionFlag.isproperty;
+            else funcFlags &= ~FunctionFlag.isproperty;
+        }
+
+        /// set or get if the function has the `ref` attribute
+        bool isref() const pure
+        {
+            return (funcFlags & FunctionFlag.isref) != 0;
+        }
+        /// ditto
+        void isref(bool v)
+        {
+            if (v) funcFlags |= FunctionFlag.isref;
+            else funcFlags &= ~FunctionFlag.isref;
+        }
+
+        /// set or get if the function has the `return` attribute
+        bool isreturn() const pure
+        {
+            return (funcFlags & FunctionFlag.isreturn) != 0;
+        }
+        /// ditto
+        void isreturn(bool v)
+        {
+            if (v) funcFlags |= FunctionFlag.isreturn;
+            else funcFlags &= ~FunctionFlag.isreturn;
+        }
+
+        /// set or get if the function has the `scope` attribute
+        bool isthisscope() const pure
```

Comment:
Why add `this`?

Reply:
Good question. It's because `Type.isscope()` already exists and has a complete different meaning. That was somewhat the big problem when writing this PR...

Reply:
That's unfortunate. Not a big fan of `this` but can't think of a better alternative right now.

Reply:
yeah. The worst is that the Type.isscope is mostly a noop but for classes.